File: tools/hikyuu/trade_sys/profitgoal.py
#!/usr/bin/python
# -*- coding: utf8 -*-
# cp936
#
# The MIT License (MIT)
#
# Copyright (c) 2010-2017 fasiondog
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.


from ._trade_sys import (ProfitGoalBase, 
                         PG_NoGoal, 
                         PG_FixedPercent,
                         PG_FixedHoldDays)
from hikyuu.util.unicode import (unicodeFunc, reprFunc)


ProfitGoalBase.__unicode__ = unicodeFunc
ProfitGoalBase.__repr__ = reprFunc


def pg_init(self, name, params):
    super(self.__class__, self).__init__(name)
    self._name = name
    self._params = params
    for k,v in params.items():
        self.setParam(k, v)
    
def crtPG(func, params={}, name='crtPG'):
    """
    快速创建自定义不带私有属性的盈利目标策略
    
    :param func: 盈利目标策略函数
    :param {} params: 参数字典
    :param str name: 自定义名称
    :return: 盈利目标策略实例
    """
    meta_x = type(name, (ProfitGoalBase,), {'__init__': pg_init})
    meta_x._clone = lambda self: meta_x(self._name, self._params)
    meta_x._calculate = func
    return meta_x(name, params)


#------------------------------------------------------------------
# add doc-string
#------------------------------------------------------------------

ProfitGoalBase.__doc__ = """
盈利目标策略基类

自定义盈利目标策略接口：

    ProfitGoalBase.getGoal() - 【必须】获取目标价格
    ProfitGoalBase._calculate() - 【必须】子类计算接口
    ProfitGoalBase._clone() - 【必须】克隆接口
    ProfitGoalBase._reset() - 【可选】重载私有变量
"""

ProfitGoalBase.name.__doc__ = """名称"""

ProfitGoalBase.__init__.__doc__ = """
__init__(self[, name="ProfitGoalBase"])
    
    初始化构造函数
        
    :param str name: 名称
"""

ProfitGoalBase.getParam.__doc__ = """
getParam(self, name)

    获取指定的参数
    
    :param str name: 参数名称
    :return: 参数值
    :raises out_of_range: 无此参数
"""

ProfitGoalBase.setParam.__doc__ = """
setParam(self, name, value)
    
    设置参数
        
    :param str name: 参数名称
    :param value: 参数值
    :type value: int | bool | float | string
    :raises logic_error: Unsupported type! 不支持的参数类型
"""

ProfitGoalBase.setTO.__doc__ = """
setTO(self, k)
    
    :param KData k: 设置交易对象
"""

ProfitGoalBase.getTO.__doc__ = """
getTO(self)
    
    :return: 交易对象
    :rtype: KData
"""

ProfitGoalBase.setTM.__doc__ = """
setTM(self, tm)
    
    :param TradeManager tm: 设置交易管理账户
"""

ProfitGoalBase.getTM.__doc__ = """
getTM(self)
    
    获取交易管理账户
        
    :rtype: TradeManager
"""

ProfitGoalBase.getGoal.__doc__ = """
getGoal(self, datetime, price)
    
    【重载接口】获取盈利目标价格，返回constant.null_price时，表示未限定目标；
    返回0意味着需要卖出。
        
    :param Datetime datetime: 买入时间
    :param float price: 买入价格
    :return: 目标价格
    :rtype: float
"""

ProfitGoalBase.reset.__doc__ = """
reset(self)
    
    复位操作
"""

ProfitGoalBase.clone.__doc__ = """
clone(self)
    
    克隆操作
"""

ProfitGoalBase._calculate.__doc__ = """
_calculate(self)
    
    【重载接口】子类计算接口
"""

ProfitGoalBase._reset.__doc__ = """
_reset(self)
    
    【重载接口】子类复位接口，复位内部私有变量
"""

ProfitGoalBase._clone.__doc__ = """
_clone(self)
    
    【重载接口】子类克隆接口
"""

ProfitGoalBase.buyNotify.__doc__ = """
buyNotify(self, trade_record)
    
    【重载接口】交易系统发生实际买入操作时，通知交易变化情况，一般存在多次增减仓的情况才需要重载
        
    :param TradeRecord trade_record: 发生实际买入时的实际买入交易记录
"""

ProfitGoalBase.sellNotify.__doc__ = """
sellNotify(self, trade_record)
    
    【重载接口】交易系统发生实际卖出操作时，通知实际交易变化情况，一般存在多次增减仓的情况才需要重载
        
    :param TradeRecord trade_record: 发生实际卖出时的实际卖出交易记录
"""


#------------------------------------------------------------------
# add doc-string for build_in func
#------------------------------------------------------------------

PG_NoGoal.__doc__ = """
PG_NoGoal()

    无盈利目标策略，通常为了进行测试或对比。
    
    :return: 盈利目标策略实例
"""

PG_FixedPercent.__doc__ = """
PG_FixedPercent([p = 0.2])

    固定百分比盈利目标，目标价格 = 买入价格 * (1 + p)
    
    :param float p: 百分比
    :return: 盈利目标策略实例
"""

PG_FixedHoldDays.__doc__ = """
PG_FixedHoldDays([days=5])

    固定持仓天数盈利目标策略
    
    :param int days: 允许持仓天数（按交易日算）,默认5天
    :return: 盈利目标策略实例
"""

File: libs/hikyuu_python/trade_sys/_ProfitGoal.cpp
/*
 * _ProfitGoal.cpp
 *
 *  Created on: 2013-3-21
 *      Author: fasiondog
 */

#include <boost/python.hpp>
#include <hikyuu/trade_sys/profitgoal/build_in.h>
#include "../_Parameter.h"
#include "../pickle_support.h"

using namespace boost::python;
using namespace hku;

class ProfitGoalWrap : public ProfitGoalBase, public wrapper<ProfitGoalBase> {
public:
    ProfitGoalWrap(): ProfitGoalBase() {}
    ProfitGoalWrap(const string& name): ProfitGoalBase(name) {}
    virtual ~ProfitGoalWrap() {}

    void _reset() {
        if (override func = get_override("_reset")) {
            func();
        } else {
            ProfitGoalBase::_reset();
        }
    }

    void default_reset() {
        this->ProfitGoalBase::_reset();
    }

    ProfitGoalPtr _clone() {
        return this->get_override("_clone")();
    }

    void _calculate() {
        this->get_override("_calculate")();
    }

    void buyNotify(const TradeRecord& tr) {
        if (override buy_notify = this->get_override("buyNotify")) {
            buy_notify(tr);
            return;
        }

        this->ProfitGoalBase::buyNotify(tr);
    }

    void default_buyNotify(const TradeRecord& tr) {
        this->ProfitGoalBase::buyNotify(tr);
    }

    void sellNotify(const TradeRecord& tr) {
        if (override sell_notify = this->get_override("sellNotify")) {
            sell_notify(tr);
            return;
        }

        this->ProfitGoalBase::sellNotify(tr);
    }

    void default_sellNotify(const TradeRecord& tr) {
        this->ProfitGoalBase::sellNotify(tr);
    }

    price_t getGoal(const Datetime& datetime, price_t price) {
        return this->get_override("getGoal")(datetime, price);
    }

    price_t getShortGoal(const Datetime& datetime, price_t price) {
        if (override getShortGoal = get_override("getShortGoal")) {
            return getShortGoal(datetime, price);
        }
        return ProfitGoalBase::getShortGoal(datetime, price);
    }

    price_t default_getShortGoal(const Datetime& datetime, price_t price) {
        return this->ProfitGoalBase::getShortGoal(datetime, price);
    }
};

string (ProfitGoalBase::*pg_get_name)() const = &ProfitGoalBase::name;
void (ProfitGoalBase::*pg_set_name)(const string&) = &ProfitGoalBase::name;

void export_ProfitGoal() {
    class_<ProfitGoalWrap, boost::noncopyable>("ProfitGoalBase", init<>())
            .def(init<const string&>())
            .def(self_ns::str(self))
            .add_property("name", pg_get_name, pg_set_name)
            .def("getParam", &ProfitGoalBase::getParam<boost::any>)
            .def("setParam", &ProfitGoalBase::setParam<object>)
            .def("setTM", &ProfitGoalBase::setTM)
            .def("setTO", &ProfitGoalBase::setTO)
            .def("getTM", &ProfitGoalBase::getTM)
            .def("getTO", &ProfitGoalBase::getTO)
            .def("buyNotify", &ProfitGoalBase::buyNotify,
                              &ProfitGoalWrap::default_buyNotify)
            .def("sellNotify", &ProfitGoalBase::sellNotify,
                               &ProfitGoalWrap::default_sellNotify)
            .def("getGoal", pure_virtual(&ProfitGoalBase::getGoal))
            .def("getShortGoal", &ProfitGoalBase::getShortGoal,
                    & ProfitGoalWrap::default_getShortGoal)
            .def("reset", &ProfitGoalBase::reset)
            .def("clone", &ProfitGoalBase::clone)
            .def("_calculate", pure_virtual(&ProfitGoalBase::_calculate))
            .def("_reset", &ProfitGoalBase::_reset, &ProfitGoalWrap::default_reset)
            .def("_clone", pure_virtual(&ProfitGoalBase::_clone))
            ;
    register_ptr_to_python<ProfitGoalPtr>();

    def("PG_NoGoal", PG_NoGoal);
    def("PG_FixedPercent", PG_FixedPercent, (arg("p") = 0.2));
    def("PG_FixedHoldDays", PG_FixedHoldDays, (arg("days") = 5));
}





